clients/admin-ui/src/features/common/custom-fields/hooks.ts

```diff
@@ -0,0 +1,191 @@
+    {
+      skip: queryFidesKey !== "" && !(isEnabled && queryFidesKey),
+    }
+  );
+
+  // The `fixedCacheKey` options will ensure that components referencing the same resource will
+  // share mutation info. That won't be too useful, though, because `upsertCustomField` can issue
+  // multiple requests: one for each field associated with the resource.
+  const [upsertCustomFieldMutationTrigger, upsertCustomFieldMutationResult] =
+    useUpsertCustomFieldMutation({ fixedCacheKey: resourceFidesKey });
+  const [deleteCustomFieldMutationTrigger, deleteCustomFieldMutationResult] =
+    useDeleteCustomFieldMutation({ fixedCacheKey: resourceFidesKey });
+
+  const isLoading =
+    allAllowListQuery.isLoading ||
+    customFieldDefinitionsQuery.isLoading ||
+    customFieldsQuery.isLoading ||
+    upsertCustomFieldMutationResult.isLoading ||
+    deleteCustomFieldMutationResult.isLoading;
+
+  const idToAllowListWithOptions = useMemo(
+    () =>
+      new Map(
+        filterWithId(allAllowListQuery.data).map((allowList) => [
+          allowList.id,
+          {
+            ...allowList,
+            options: (allowList.allowed_values ?? []).map((value) => ({
+              value,
+              label: value,
+            })),
+          },
+        ])
+      ),
+    [allAllowListQuery.data]
+  );
+
+  const idToCustomFieldDefinition = useMemo(
+    () =>
+      new Map(
+        filterWithId(customFieldDefinitionsQuery.data).map((cfd) => [
+          cfd.id,
+          cfd,
+        ])
+      ),
+    [customFieldDefinitionsQuery]
+  );
+
+  const definitionIdToCustomField = useMemo(
+    () =>
+      new Map(
+        filterWithId(customFieldsQuery.data).map((fd) => [
+          fd.custom_field_definition_id,
+          fd,
+        ])
+      ),
+    [customFieldsQuery]
+  );
+
+  const sortedCustomFieldDefinitionIds = useMemo(() => {
+    const ids = [...idToCustomFieldDefinition.keys()];
+    ids.sort();
+    return ids;
+  }, [idToCustomFieldDefinition]);
+
+  /**
+   * Issue a batch of upsert and delete requests that will sync the form selections to the
+   * backend.
+   */
+  const upsertCustomFields = useCallback(
+    async (formValues: CustomFieldsFormValues) => {
+      // When creating an resource, the fides key may have initially been blank. But by the time the
+      // form is submitted it must not be blank (not undefined, not an empty string).
+      const fidesKey = formValues.fides_key || resourceFidesKey;
+      if (!fidesKey) {
+        return;
+      }
+
+      const { definitionIdToCustomFieldValue } = formValues;
+
+      // This will be undefined if the form never rendered a `CustomFieldList` that would assign
+      // form values.
+      if (!definitionIdToCustomFieldValue) {
+        return;
+      }
+
+      try {
+        // This would be a lot simpler (and more efficient) if the API had an endpoint for updating
+        // all the metadata associated with a field, including deleting options that weren't passed.
```

Comment:
Should I create an issue for this?

Reply:
You are referencing your comment on line 130?

Reply:
https://github.com/ethyca/fides/issues/2612